File: src/main/java/seremis/geninfusion/soul/gene/newAI/GeneMoveSpeed.java
package seremis.geninfusion.soul.gene.newAI;

import seremis.geninfusion.api.soul.IChromosome;
import seremis.geninfusion.soul.Gene;
import seremis.geninfusion.soul.allele.AlleleDouble;
import seremis.geninfusion.soul.allele.AlleleDoubleArray;

public abstract class GeneMoveSpeed extends Gene {

    @Override
    public IChromosome mutate(IChromosome chromosome) {
        if(possibleAlleles().equals(AlleleDouble.class)) {
            AlleleDouble allele1 = (AlleleDouble) chromosome.getPrimary();
            AlleleDouble allele2 = (AlleleDouble) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                allele1.value = allele1.value * ((rand.nextFloat() * 2 * 0.1) + 0.9);
            } else {
                allele2.value = allele2.value * ((rand.nextFloat() * 2 * 0.1) + 0.9);
            }
        } else if(possibleAlleles().equals(AlleleDoubleArray.class)) {
            AlleleDoubleArray allele1 = (AlleleDoubleArray) chromosome.getPrimary();
            AlleleDoubleArray allele2 = (AlleleDoubleArray) chromosome.getSecondary();

            for(int i = 0; i < allele1.value.length; i++) {
                if(rand.nextBoolean()) {
                    allele1.value[i] = allele1.value[i] * ((rand.nextFloat() * 2 * 0.1) + 0.9);
                } else {
                    allele2.value[i] = allele2.value[i] * ((rand.nextFloat() * 2 * 0.1) + 0.9);
                }
            }
        }
        return chromosome;
    }
}

File: src/main/java/seremis/geninfusion/soul/gene/GeneMoveSpeed.scala
package seremis.geninfusion.soul.gene

import seremis.geninfusion.api.soul.{EnumAlleleType, IChromosome}
import seremis.geninfusion.soul.Gene

class GeneMoveSpeed(alleleType: EnumAlleleType) extends Gene(alleleType) {

    override def mutate(chromosome: IChromosome): IChromosome = {
        if(mutate) {
            var allele1Data = chromosome.getPrimary.getAlleleData
            var allele2Data = chromosome.getSecondary.getAlleleData

            alleleType match {
                case EnumAlleleType.DOUBLE =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Double] * ((rand.nextFloat() * 2 * 0.1) + 0.9)
                    else
                        allele2Data = allele2Data.asInstanceOf[Double] * ((rand.nextFloat() * 2 * 0.1) + 0.9)
                case EnumAlleleType.DOUBLE_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Double]].length)
                        allele1Data.asInstanceOf[Array[Double]](index) = allele1Data.asInstanceOf[Array[Double]](index) * ((rand.nextFloat() * 2 * 0.1) + 0.9)
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Double]].length)
                        allele2Data.asInstanceOf[Array[Double]](index) = allele2Data.asInstanceOf[Array[Double]](index) * ((rand.nextFloat() * 2 * 0.1) + 0.9)
                    }
            }
        }
        chromosome
    }
}
